pkg/maps/egressmap: refactor into a cell

lmb opened the pull request:
pkg/maps/egressmap: refactor policy map into a cell

    Rework the egressmap from a global variable into a Cell. The map still has
    singleton behaviour due to map pinning, but at least for unit tests we can
    opt out of pinning and isolate the map from global changes.

    Stop writing out EGRESS_POLICY_MAP_SIZE into node_config.h since the map
    definition in C is never actually used. The map size is configured via a
    command line parameter instead. Hardcode the previously used default value
    to avoid errors when running BPF unit tests. This is simpler than wiring the
    new PolicyConfig into HeaderfileWriter.

    Reuse the MapOut strategy from configmap and authmap to ensure that the
    egressmap is initialized by the agent before the loader is invoked. 
    Otherwise map creation might use the (incorrect) MaxElems from the compiled
    C code.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

hivetest: add Lifecycle to make testing easier

    hive requires splitting up object lifecycle into New, Start, Stop. It does
    this by injecting a hive.Lifecycle into constructors. This in turn means
    that we need a lifecycle to create objects during testing.

    Implement a minimal hive.Lifecycle for use in testing, which doesn't 
    distinguish between New and Start phases and instead immediately executes
    any start hooks. Stop hooks are invoked at test cleanup time.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>


```release-note
Refactor egressgateway specific maps into a cell
```

Updates: #23782

lmb commented on pkg/egressgateway/policy.go:
```diff
@@ -17,11 +17,20 @@ import (
 )
 
+// PolicyMap is used to communicate EGW policies to the datapath.
+type PolicyMap interface {
```
This could live in `egressmap`, but then we hide the concrete implementation of the policy map by default. That gets annoying if some other part of the code base needs another method, which then forces egressgw mocks / fakes to implement that method as well.

Happy to move this around if someone feels strongly about it.

lmb replied:
Ended up moving this, testing becomes a mess otherwise...

asauber commented on pkg/maps/egressmap/policy.go:
```diff
@@ -33,41 +37,66 @@ type EgressPolicyVal4 struct {
-			MaxEntries: uint32(maxPolicyEntries),
-			Pinning:    ebpf.PinByName,
-		})
+func (def PolicyConfig) Flags(flags *pflag.FlagSet) {
```
Can you point me to some Hive code or documentation which indicates how this Flag is initialized (where this function is invoked). I guess it has something to do with this being a method of [`cell.Config(DefaultPolicyConfig)` above](https://github.com/cilium/cilium/pull/24865/files/ce64e65f64968d77457d190dbaeb121268640fa4#diff-afe681b1a0a32f6119e18f7484f3d640f1f7fffbd798dfcd490e19db72ae9494R43).

It seems strange to bury a user-facing flag in this file, but maybe that's the new pattern!

lmb replied:
Not sure how initialization happens tbh, it is the canonical way to to things with hive though. Here is the example, maybe that helps? https://github.com/cilium/cilium/blob/v1.13.2/pkg/hive/example/main.go

asauber commented on cilium/cmd/bpf_egress_list.go:
```diff
@@ -38,7 +38,8 @@ var bpfEgressListCmd = &cobra.Command{
 		common.RequireRootPrivilege("cilium bpf egress list")
 
-		if err := egressmap.OpenEgressMaps(); err != nil {
+		policyMap, err := egressmap.OpenPinnedPolicyMap()
```
Rather than call a function to return the singleton provided by the Hive cell, is there not some way to use the new Cell to provide this Map as a dependency via the Hive graph?

lmb replied:
Maybe, but I'm not sure how to tie hive into this command. As I understand it, we currently do `agent cobra command -> create hive -> execute agent`. This is in a separate cobra command. Also note that here we don't want to create the map, we only want to open it if it already exists. The Cell will actually create the map as well.

mhofstetter commented on pkg/maps/egressmap/policy.go:
```diff
@@ -33,41 +37,66 @@ type EgressPolicyVal4 struct {
+		Pinning:    pinning,
+	})
+
+	if err := m.OpenOrCreate(); err != nil {
```
AFAIK we should keep the constructors side-effect free - therefore moving the call to `OpenOrCreate` into a start hook

lmb commented on bpf/lib/maps.h:
```diff
@@ -213,7 +213,7 @@ struct {
 	__type(value, struct egress_gw_policy_entry);
 	__uint(pinning, LIBBPF_PIN_BY_NAME);
-	__uint(max_entries, EGRESS_POLICY_MAP_SIZE);
+	__uint(max_entries, 16384);
```
This is to keep the BPF unit tests happy.

viktor-kurchenko replied:
So, value must be hardcoded and constant is not allowed?

lmb replied:
I removed the define for `EGRESS_POLICY_MAP_SIZE` so that I can get rid of the entry in `node_config.h` which is written out at runtime. It's difficult to plumb the size requested via config / command line into `node_config.h` without relying on `option.Config` global which I wanted to avoid.

So: we could keep the `define` but move it to `maps.h`. That runs the risk of some C code relying on that define with the assumption that it reflects the configured value. So I opted to remove the define instead.

viktor-kurchenko replied:
Thanks!

lmb committed:
hivetest: add Lifecycle to make testing easier

hive requires splitting up object lifecycle into New, Start, Stop.
It does this by injecting a hive.Lifecycle into constructors. This
in turn means that we need a lifecycle to create objects during testing.

Implement a minimal hive.Lifecycle for use in testing, which doesn't
distinguish between New and Start phases and instead immediately
executes any start hooks. Stop hooks are invoked at test cleanup time.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb commented on pkg/maps/egressmap/policy.go:
```diff
@@ -33,41 +39,74 @@ type EgressPolicyVal4 struct {
+	return bpf.NewMapOut(PolicyMap(CreatePolicyMap(lc, cfg)))
+}
+
+func CreatePolicyMap(lc hive.Lifecycle, cfg PolicyConfig) *policyMap {
```
`PolicyMap`

lmb commented on pkg/maps/egressmap/policy.go:
```diff
@@ -33,41 +39,74 @@ type EgressPolicyVal4 struct {
-			return err
-		}
+func createPolicyMapFromDaemonConfig(daemonConfig *option.DaemonConfig, lc hive.Lifecycle, cfg PolicyConfig) bpf.MapOut[PolicyMap] {
+	if !daemonConfig.EnableIPv4EgressGateway {
```
Maybe we can get away without this? Ties into the whole "optional hive feature".

jibi commented on pkg/datapath/cells.go:
```diff
@@ -101,6 +101,7 @@ type datapathParams struct {
 	WgAgent *wg.Agent
 
-	// Force map initialisation before loader
+	// Force map initialisation before loader. You should not use these otherwise.
```
nit: extra dot at the end

lmb replied:
Hmm I don't understand, sorry.

jibi commented on pkg/egressgateway/manager_test.go:
```diff
@@ -0,0 +1,17 @@
+	"github.com/cilium/cilium/pkg/hive"
+)
+
+func TestCell(t *testing.T) {
```
why is this in a separated from `manager_privileged_test.go`?

lmb replied:
Because it's not privileged. I can move it into the other file if you prefer.

jibi reviewed (approved):
Looks good, thanks! Just one nit and a question

mhofstetter reviewed (approved):
LGTM - thanks for introducing the generic `MapOut` type

lmb committed:
pkg/maps/egressmap: refactor policy map into a cell

Rework the egressmap from a global variable into a Cell. The map still
has singleton behaviour due to map pinning, but at least for unit tests
we can opt out of pinning and isolate the map from global changes.

Stop writing out EGRESS_POLICY_MAP_SIZE into node_config.h since
the map definition in C is never actually used. The map size is
configured via a command line parameter instead. Hardcode the previously
used default value to avoid errors when running BPF unit tests.
This is simpler than wiring the new PolicyConfig into HeaderfileWriter.

Reuse the MapOut strategy from configmap and authmap to ensure that
the egressmap is initialized by the agent before the loader is invoked.
Otherwise map creation might use the (incorrect) MaxElems from the
compiled C code.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

joamaki reviewed (approved):
hivetest package looks good to me

thorn3r reviewed (approved):
LGTM, this was a nice lil intro to hive for me :sunglasses: nice work
